daemons/fenced/fenced_commands.c

```diff
@@ -3026,297 +2988,429 @@ is_privileged(pcmk__client_t *c, const char *op)
-        reply = fenced_construct_reply(request->xml, NULL, &request->result);
+// CRM_OP_RM_NODE_CACHE
+static xmlNode *
+handle_cache_request(pcmk__request_t *request)
+{
+    int node_id = 0;
+    const char *name = NULL;
 
-    } else if (pcmk__str_eq(op, STONITH_OP_LEVEL_ADD, pcmk__str_none)) {
-        char *device_id = NULL;
+    crm_element_value_int(request->xml, XML_ATTR_ID, &node_id);
+    name = crm_element_value(request->xml, XML_ATTR_UNAME);
+    reap_crm_member(node_id, name);
+    pcmk__set_result(&request->result, CRM_EX_OK, PCMK_EXEC_DONE, NULL);
+    return NULL;
+}
 
-        if (is_privileged(request->client, op)) {
-            fenced_register_level(request->xml, &device_id, &request->result);
-        } else {
-            pcmk__set_result(&request->result, CRM_EX_INSUFFICIENT_PRIV,
-                             PCMK_EXEC_INVALID,
-                             "Unprivileged users must add level via CIB");
-        }
-        fenced_send_level_notification(op, &request->result, device_id);
-        free(device_id);
-        reply = fenced_construct_reply(request->xml, NULL, &request->result);
+static xmlNode *
+handle_unknown_request(pcmk__request_t *request)
+{
+    const char *op = crm_element_value(request->xml, F_STONITH_OPERATION);
 
-    } else if (pcmk__str_eq(op, STONITH_OP_LEVEL_DEL, pcmk__str_none)) {
-        char *device_id = NULL;
+    crm_err("Unknown IPC request %s from %s %s",
+            op, pcmk__request_origin_type(request),
+            pcmk__request_origin(request));
+    pcmk__format_result(&request->result, CRM_EX_PROTOCOL, PCMK_EXEC_INVALID,
+                        "Unknown IPC request type '%s' (bug?)", crm_str(op));
+    return fenced_construct_reply(request->xml, NULL, &request->result);
+}
 
-        if (is_privileged(request->client, op)) {
-            fenced_unregister_level(request->xml, &device_id, &request->result);
-        } else {
-            pcmk__set_result(&request->result, CRM_EX_INSUFFICIENT_PRIV,
-                             PCMK_EXEC_INVALID,
-                             "Unprivileged users must delete level via CIB");
-        }
-        fenced_send_level_notification(op, &request->result, device_id);
-        reply = fenced_construct_reply(request->xml, NULL, &request->result);
+static void
+handle_request(pcmk__request_t *request)
+{
+    xmlNode *reply = NULL;
+    const char *op = NULL;
+    const char *reason = NULL;
 
-    } else if(pcmk__str_eq(op, CRM_OP_RM_NODE_CACHE, pcmk__str_casei)) {
-        int node_id = 0;
-        const char *name = NULL;
+    CRM_CHECK((request != NULL) && (request->xml != NULL), return);
 
-        crm_element_value_int(request->xml, XML_ATTR_ID, &node_id);
-        name = crm_element_value(request->xml, XML_ATTR_UNAME);
-        reap_crm_member(node_id, name);
-        pcmk__set_result(&request->result, CRM_EX_OK, PCMK_EXEC_DONE, NULL);
+    op = crm_element_value(request->xml, F_STONITH_OPERATION);
```

Comment:
we might save a little bit of xml-parsing if we memorize the op in the request structure.